python/ray/tune/syncer.py

```diff
@@ -142,395 +79,408 @@ class SyncConfig:
-        # String or callable (for function syncers)
-        sync_client = get_sync_client(sync_function)
-    else:
-        # sync_function=True or sync_function=None --> default
-        rsync_function_str = get_rsync_template_if_available()
-        if rsync_function_str:
-            sync_client = CommandBasedClient(rsync_function_str, rsync_function_str)
-            sync_client.set_logdir(local_dir)
-        else:
-            sync_client = RemoteTaskClient()
+    def _local_trial_logdir(self, trial: "Trial"):
+        return trial.logdir
 
-    _syncers[key] = NodeSyncer(local_dir, remote_dir, sync_client)
-    return _syncers[key]
+    def _remote_trial_logdir(self, trial: "Trial"):
+        return trial.logdir
 
+    def _sync_trial_dir(
+        self, trial: "Trial", force: bool = False, wait: bool = True
+    ) -> bool:
+        if not self._enabled or trial.uses_cloud_checkpointing:
+            return False
 
-@DeveloperAPI
-class SyncerCallback(Callback):
-    def __init__(self, sync_function: Optional[Union[bool, Callable]]):
-        self._sync_function = sync_function
-        self._syncers: Dict["Trial", NodeSyncer] = {}
-
-    def _get_trial_syncer(self, trial: "Trial"):
-        if trial not in self._syncers:
-            self._syncers[trial] = self._create_trial_syncer(trial)
-        return self._syncers[trial]
-
-    def _create_trial_syncer(self, trial: "Trial"):
-        return get_node_syncer(
-            trial.logdir, remote_dir=trial.logdir, sync_function=self._sync_function
-        )
+        sync_process = self._get_trial_sync_process(trial)
 
-    def _remove_trial_syncer(self, trial: "Trial"):
-        self._syncers.pop(trial, None)
+        if not force and not self._should_sync(trial):
+            return False
 
-    def _sync_trial_checkpoint(self, trial: "Trial", checkpoint: _TrackedCheckpoint):
-        if checkpoint.storage_mode == CheckpointStorage.MEMORY:
-            return
+        if NODE_IP in trial.last_result:
+            source_ip = trial.last_result[NODE_IP]
+        else:
+            source_ip = ray.get(trial.runner.get_current_ip.remote())
 
-        trial_syncer = self._get_trial_syncer(trial)
-        # If the sync_function is False, syncing to driver is disabled.
-        # In every other case (valid values include None, True Callable,
-        # NodeSyncer) syncing to driver is enabled.
-        if trial.sync_on_checkpoint and self._sync_function is not False:
+        try:
+            sync_process.wait()
+        except TuneError as e:
+            # Errors occurring during this wait are not fatal for this
+            # checkpoint, so it should just be logged.
+            logger.error(
+                f"Trial {trial}: An error occurred during the "
+                f"checkpoint syncing of the previous checkpoint: {e}"
+            )
+        sync_process.start(
+            source_ip=source_ip,
+            source_path=self._remote_trial_logdir(trial),
```

Comment:
do you know why we are syncing the whole trial logdir instead of this particular checkpoint_00000x folder?

Reply:
Can you point me to the code?